Print out snabb config get output in XPath format

dpino opened the pull request:
Fixed #837 . I'm not sure whether this is the best approach. Basically it transforms a standard output to an xpath output. 

The PR doesn't tackle Mislav's suggest to print out default values (this could be a different issue).

Example:

```
$ sudo ./snabb config get 2419 /softwire-config/external-interface
allow-incoming-icmp false;
error-rate-limiting {
  packets 600000;
}
generate-icmp-errors false;
ip 10.0.1.1;
mac 02:aa:aa:aa:aa:aa;
next-hop {
  mac 02:99:99:99:99:99;
}
reassembly {
  max-fragments-per-packet 40;
}
```

```bash
$ sudo ./snabb config get --xpath-format 2419 /softwire-config/external-interface                                                                    
/allow-incoming-icmp false
/error-rate-limiting/packets 600000
/generate-icmp-errors false
/ip 10.0.1.1
/mac 02:aa:aa:aa:aa:aa
/next-hop/mac 02:99:99:99:99:99
/reassembly/max-fragments-per-packet 40
```

mislavn commented:
Thank you @dpino.

@wingo is it possible to add a flag to snabb config so it shows the default values also?

mislavn commented:
@dpino can you also implement this for state data?

dpino commented:
@mislavn yep, no problem, I will send a new commit for get-state. Default variables is also doable. Andy and I discussed it offline and most likely we will add a knob to force print out of default variables (we think it's still valuable in some cases to not print out the defaults to reduce output size).

dpino commented:
xpath-format in get-state example:

```
$ sudo ./snabb config get-state --xpath-format 11973 /
/softwire-state/drop-all-ipv4-iface-bytes 0
/softwire-state/drop-all-ipv4-iface-packets 0
/softwire-state/drop-all-ipv6-iface-bytes 0
/softwire-state/drop-all-ipv6-iface-packets 0
/softwire-state/drop-bad-checksum-icmpv4-bytes 0
...
/softwire-state/out-ipv6-frag-not 0
/softwire-state/out-ipv6-packets 0
```

dpino commented:
Added a kbob to print out default values. Example:

**Run lwAFTR**

```bash
$ sudo ./snabb lwaftr run --name lwaftr --reconfigurable --conf lwaftr-migrated-2.conf --on-a-stick 82:00.1
```

**Query standard external-interface**

```bash
$ sudo ./snabb config get lwaftr /softwire-config/external-interface
allow-incoming-icmp false;
error-rate-limiting {
  packets 600000;
}
generate-icmp-errors false;
ip 10.0.1.1;
mac 02:aa:aa:aa:aa:aa;
next-hop {
  mac 02:99:99:99:99:99;
}
reassembly {
  max-fragments-per-packet 40;
}
```

**Query external-interface printing out default values**

```bash
$ sudo ./snabb config get --print-default lwaftr /softwire-config/external-interface
allow-incoming-icmp false;
error-rate-limiting {
  packets 600000;
  period 2;
}
generate-icmp-errors false;
ip 10.0.1.1;
mac 02:aa:aa:aa:aa:aa;
mtu 1460;
next-hop {
  mac 02:99:99:99:99:99;
}
reassembly {
  max-fragments-per-packet 40;
  max-packets 20000;
}
```

wingo commented on src/apps/config/leader.lua:
```diff
@@ -114,28 +114,30 @@ function Leader:rpc_describe (args)
-local function path_printer_for_grammar(grammar, path)
+local function path_printer_for_grammar(grammar, path, print_default)
    local getter, subgrammar = path_mod.resolver(grammar, path)
+   data.print_default(print_default or false)
```
"print_default or false" is the same as "print_default", at least in a boolean context.

dpino replied:
Right, but there's a chance that print_default is nil when launching `config get` or `config get-state` without --print-default knob. I can force print_default to be false by default though.

wingo commented on src/lib/yang/data.lua:
```diff
@@ -803,11 +803,19 @@ function data_printer_from_grammar(production)
       return function(data, file, indent)
          local str = serialize(data)
-         if str ~= production.default then
+         if print_default() or str ~= production.default then
```
Here print_default() always returns a true value AFAIU -- it returns the function from line 808.  Anyway this doesn't look right to me: can you change instead to add a "print_default" option on line 707?  The idea being, you derive a printer from a grammar either with default-printing or without.  It's a property of the printer.  That way you can just check a simple variable in the scalar printer to know whether to print defaults or no.

wingo commented on src/program/config/xpath_printer.lua:
```diff
@@ -0,0 +1,97 @@
+      local tok = p:token()
+      if tok == '}' then
+         table.remove(path)
+      elseif p:peep() == '{' then
```
This isn't going to work in general.  The format for the xpath has one component per tree, yes, but for lists for example you are going to have e.g. `/foo/bar[a=A][b=B]/qux'.  I think doing this transformation on the client side is the wrong thing; we need to tell the leader that we want output in xpath format and have it generate it, using a printer derived from the grammar.  That way we also avoid paging the whole thing into memory on the client side.

wingo reviewed (changes requested):
For the future, I think this PR would have been better if it were just one of these two features.  That said, it's a good start.  Some changes requested for both features.

dpino committed:
Update docs

dpino committed:
Implement XPath data printer

dpino committed:
Support xpath-printer knob

dpino commented:
Printing of default-values was refactored out from this PR and moved to https://github.com/Igalia/snabb/pull/850.

I reworked the whole code and implemented an XPath printer similar to the standard data_printer, as Andy suggested. It's still work in progress, although the example above is working OK. I'd need to think more about how to handle arrays, tables and other data types.

Example:

```
$ sudo ./snabb lwaftr run --reconfigurable --name lwaftr2 --conf lwaftr.conf --on-a-stick 82:00.0
```

```
$ sudo ./snabb config get --xpath-format lwaftr2 /softwire-config/external-interface
/softwire-config/external-interface/allow-incoming-icmp false;
/softwire-config/external-interface/error-rate-limiting/packets 600000;
/softwire-config/external-interface/ip 10.10.10.10;
/softwire-config/external-interface/mac 12:12:12:12:12:12;
/softwire-config/external-interface/next-hop/mac 68:68:68:68:68:68;
/softwire-config/external-interface/reassembly/max-fragments-per-packet 40;
```

wingo commented on src/lib/yang/snabb-config-leader-v1.yang:
```diff
@@ -39,6 +39,7 @@ module snabb-config-leader-v1 {
       leaf revision { type string; }
       leaf path { type string; default "/"; }
       leaf print-default { type boolean; }
+      leaf xpath-format { type boolean; }
```
can this option be "format=xpath" with the default being "format=yang"

wingo commented on src/program/config/README.md:
````diff
@@ -135,6 +135,19 @@ In the example above, attributes sucha as `period` and `mtu` take their
+is selected, every attribute is printed as an absolute path.  Example:
+
+```
+$ sudo ./snabb config get --xpath-format lwaftr2 /softwire-config/external-interface
````
Likewise can the command-line argument be `--format=xpath`

wingo commented:
For me, generally looking good; need to change to correctly print xpath paths for lists (the key goes in the path)

dpino committed:
Rename xpath-format to format

dpino committed:
Print out table in XPath format

dpino commented:
Binding-table example (XPath format):

```
$ sudo ./snabb config get --format=xpath lwaftr /softwire-config/binding-table                                                                                 
/softwire-config/binding-table/[ipv4=178.79.150.3;psid=4]/b4-ipv6 127:14:25:36:47:58:69:128;
/softwire-config/binding-table/[ipv4=178.79.150.3;psid=4]/br-address 1e:2:2:2:2:2:2:af;
/softwire-config/binding-table/[ipv4=178.79.150.3;psid=4]/psid-length 6;
/softwire-config/binding-table/[ipv4=178.79.150.233;psid=22788]/b4-ipv6 127:11:12:13:14:15:16:128;
/softwire-config/binding-table/[ipv4=178.79.150.233;psid=22788]/br-address 8:9:a:b:c:d:e:f;
/softwire-config/binding-table/[ipv4=178.79.150.233;psid=22788]/psid-length 16;
```

External-interface example (XPath format):

```
$ sudo ./snabb config get --format=xpath lwaftr /softwire-config/external-interface
/softwire-config/external-interface/allow-incoming-icmp false;
/softwire-config/external-interface/packets 600000;
/softwire-config/external-interface/ip 10.10.10.10;
/softwire-config/external-interface/mac 12:12:12:12:12:12;
/softwire-config/external-interface/mac 68:68:68:68:68:68;
/softwire-config/external-interface/max-fragments-per-packet 40;
```

dpino committed:
Fix key composition

dpino commented:
Fixed key composition in XPath format:

```
$ sudo ./snabb config get --format=xpath lwaftr /softwire-config/binding-table
/softwire-config/binding-table/softwire[ipv4=178.79.150.3][psid=4]/b4-ipv6 127:14:25:36:47:58:69:128;
/softwire-config/binding-table/softwire[ipv4=178.79.150.3][psid=4]/br-address 1e:2:2:2:2:2:2:af;
/softwire-config/binding-table/softwire[ipv4=178.79.150.3][psid=4]/psid-length 6;
/softwire-config/binding-table/softwire[ipv4=178.79.150.233][psid=22788]/b4-ipv6 127:11:12:13:14:15:16:128;
/softwire-config/binding-table/softwire[ipv4=178.79.150.233][psid=22788]/br-address 8:9:a:b:c:d:e:f;
/softwire-config/binding-table/softwire[ipv4=178.79.150.233][psid=22788]/psid-length 16;
/softwire-config/binding-table/softwire[ipv4=178.79.150.1][psid=0]/b4-ipv6 127:10:20:30:40:50:60:128;
/softwire-config/binding-table/softwire[ipv4=178.79.150.1][psid=0]/br-address 8:9:a:b:c:d:e:f;
/softwire-config/binding-table/softwire[ipv4=178.79.150.1][psid=0]/psid-length 0;
```

wingo commented on src/lib/yang/data.lua:
```diff
@@ -701,6 +701,205 @@ local function value_serializer(typ)
+      end
+   end
+   function handlers.array(keyword, production)
+      local serialize = value_serializer(production.element_type)
```
I think this also needs additional work (should print as [position()=1] e.g.)

dpino committed:
Handle special cases when root path includes trailing slash

dpino committed:
Support print out of arrays in XPath format

dpino committed:
Append accumulated path when print a struct keyword

dpino commented:
Implemented support for printing out arrays in XPath format. There's currently no `leaf-list` elements in snabb-softwire-v2.yang, which are compiled to arrays. In order to test the implementation was working OK, I temporarily added a leaf-list to the schema.

```
     container external-interface {
       description
        "Configuration for the external, internet-facing IPv4
@@ -168,6 +167,10 @@ module snabb-softwire-v2 {
          "Maximum packet size to send on the IPv4 interface.";
       }

+      leaf-list public-ip {
+        type inet:ipv4-address;
+      }
+
       uses traffic-filters;
```

Example:

```
$ sudo ./snabb config get --format=xpath lwaftr2 /softwire-config/external-interface
/softwire-config/external-interface/allow-incoming-icmp false;
/softwire-config/external-interface/error-rate-limiting/packets 600000;
/softwire-config/external-interface/ip 10.10.10.10;
/softwire-config/external-interface/mac 12:12:12:12:12:12;
/softwire-config/external-interface/next-hop/mac 68:68:68:68:68:68;
/softwire-config/external-interface[position()=1]/public-ip 10.10.10.10;
/softwire-config/external-interface[position()=2]/public-ip 10.10.10.11;
/softwire-config/external-interface/reassembly/max-fragments-per-packet 40;
```

dpino commented:
Mislav informed me via private conversation semicolons are not needed. As for absolute and relative paths, there's no preference. I must said here that when directly querying a leaf, there's no attribute value printed, thus there's no absolute path. For instance:

```
sudo ./snabb config get --format=xpath lwaftr2 /softwire-config/external-interface/mac
12:12:12:12:12:12;
```

Is this consistent? Should we prefer relative paths instead?

wingo commented:
On Tue 04 Jul 2017 15:17, Diego Pino <notifications@github.com> writes:

> Mislav informed me via private conversation semicolons are not needed. As for absolute and relative paths, there's no preference. I must said here that when directly querying a leaf, there's no attribute value printed, thus there's no absolute path. For instance:
>
> sudo ./snabb config get --format=xpath lwaftr2 /softwire-config/external-interface/mac
> 12:12:12:12:12:12;
>
> Is this consistent? Should we prefer relative paths instead?

I think we should prepend the queried absolute path to all results, so
that all results have full paths.  Just MHO!

dpino commented:
@wingo: Sounds reasonable. So I will tackle those changes in a new set of commits (do not print semicolon and print absolute path for queries that only return data).

dpino committed:
Top printers should print absolute path too

dpino committed:
Support print_default in scalar top_printer

dpino committed:
Remove trailing semicolon

dpino commented:
I think is working fine now. Some examples:

```
$ sudo ./snabb config get --format=xpath lwaftr /softwire-config/binding-table/softwire[ipv4=178.79.150.2][psid=7850]/br-address
/softwire-config/binding-table/softwire[ipv4=178.79.150.2][psid=7850]/br-address 1e:1:1:1:1:1:1:af
```


```
$ sudo ./snabb config get --format=xpath lwaftr /softwire-config/binding-table
/softwire-config/binding-table/softwire[ipv4=178.79.150.3][psid=4]/b4-ipv6 127:14:25:36:47:58:69:128
/softwire-config/binding-table/softwire[ipv4=178.79.150.3][psid=4]/br-address 1e:2:2:2:2:2:2:af
/softwire-config/binding-table/softwire[ipv4=178.79.150.3][psid=4]/port-set/psid-length 6
/softwire-config/binding-table/softwire[ipv4=178.79.150.233][psid=22788]/b4-ipv6 127:11:12:13:14:15:16:128
/softwire-config/binding-table/softwire[ipv4=178.79.150.233][psid=22788]/br-address 8:9:a:b:c:d:e:f
/softwire-config/binding-table/softwire[ipv4=178.79.150.233][psid=22788]/port-set/psid-length 16
```

wingo reviewed (approved):
lgtm.  thanks for this!
